pkg/controllers/node/termination/controller.go

```diff
@@ -131,23 +130,58 @@ func (c *Controller) finalize(ctx context.Context, node *corev1.Node) (reconcile
 		return reconcile.Result{}, serrors.Wrap(fmt.Errorf("tainting node, %w", err), "taint", pretty.Taint(v1.DisruptedNoScheduleTaint))
 	}
 
+	var stored *v1.NodeClaim
+	if nodeClaim != nil {
+		stored = nodeClaim.DeepCopy()
+	}
+	var result reconcile.Result
 	for _, f := range []terminationFunc{
 		c.awaitDrain,
 		c.awaitVolumeDetachment,
 		c.awaitInstanceTermination,
 	} {
-		result, err := f(ctx, nodeClaim, node, nodeTerminationTime)
-		if result != nil || err != nil {
-			return *result, err
+		result, err = f(ctx, nodeClaim, node, nodeTerminationTime)
+		if !lo.IsEmpty(result) || err != nil {
+			break
+		}
+	}
+	// If we don't have a NodeClaim, then there's nothing for us to patch here
+	if stored != nil && !equality.Semantic.DeepEqual(stored, nodeClaim) {
+		// We use client.MergeFromWithOptimisticLock because patching a list with a JSON merge patch
+		// can cause races due to the fact that it fully replaces the list on a change
+		// Here, we are updating the status condition list
+		if err = c.kubeClient.Status().Patch(ctx, nodeClaim, client.MergeFromWithOptions(stored, client.MergeFromWithOptimisticLock{})); client.IgnoreNotFound(err) != nil {
```

Comment:
This simplifies the terminationFunc code a lot so that we don't have to shove all of our checks down in there -- the trade-off is that we don't get a guarantee that this patch actually succeeds when we are tearing down until after the patch call succeeds